src/llvm/di.rs

```diff
@@ -204,198 +204,255 @@ impl DISanitizer {
+            }
+
+            // Remove retained nodes from the old program or we'll hit a debug assertion since
+            // its debug variables no longer point to the program. See the
+            // NumAbstractSubprograms assertion in DwarfDebug::endFunctionImpl in LLVM.
+            let empty_node =
+                unsafe { LLVMMDNodeInContext2(self.context, core::ptr::null_mut(), 0) };
+            sub_program.set_retained_nodes(empty_node);
+
+            let ret = replace.insert(sub_program.value_ref as u64, unsafe {
+                LLVMValueAsMetadata(new_program.value_ref)
+            });
+            assert!(ret.is_none());
+        }
 
-unsafe fn is_globalobject(v: LLVMValueRef) -> bool {
-    !LLVMIsAGlobalObject(v).is_null()
+        replace
+    }
 }
 
-unsafe fn can_get_all_metadata(v: LLVMValueRef) -> bool {
-    is_globalobject(v) || is_instruction(v)
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum Item {
+    GlobalVariable(LLVMValueRef),
+    GlobalAlias(LLVMValueRef),
+    Function(LLVMValueRef),
+    FunctionParam(LLVMValueRef),
+    Instruction(LLVMValueRef),
+    Operand(Operand),
+    MetadataEntry(LLVMValueRef, u32, usize),
 }
 
-unsafe fn can_get_operands(v: LLVMValueRef) -> bool {
-    is_mdnode(v) || is_user(v)
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct Operand {
+    parent: LLVMValueRef,
+    value: LLVMValueRef,
+    index: u32,
 }
 
-pub struct Cache {
-    keys: HashSet<u64>,
+impl Operand {
+    fn replace(&mut self, value: LLVMValueRef) {
+        unsafe {
+            if !LLVMIsAMDNode(self.parent).is_null() {
+                let value = LLVMValueAsMetadata(value);
+                LLVMReplaceMDNodeOperandWith(self.parent, self.index, value);
+            } else if !LLVMIsAUser(self.parent).is_null() {
+                LLVMSetOperand(self.parent, self.index, value);
+            }
+        }
+    }
 }
 
-impl Cache {
-    pub fn new() -> Self {
-        Cache {
-            keys: HashSet::new(),
+impl Item {
+    fn value_ref(&self) -> LLVMValueRef {
+        match self {
+            Item::GlobalVariable(value) => *value,
+            Item::GlobalAlias(value) => *value,
+            Item::Function(value) => *value,
+            Item::FunctionParam(value) => *value,
+            Item::Instruction(value) => *value,
+            Item::Operand(Operand { value, .. }) => *value,
+            Item::MetadataEntry(value, _, _) => *value,
         }
     }
 
-    pub fn hit(&mut self, key: u64) -> bool {
-        !self.keys.insert(key)
+    fn value_id(&self) -> u64 {
+        match self {
+            Item::GlobalVariable(value) => *value as u64,
+            Item::GlobalAlias(value) => *value as u64,
+            Item::Function(value) => *value as u64,
+            Item::FunctionParam(value) => *value as u64,
+            Item::Instruction(value) => *value as u64,
+            Item::Operand(Operand { value, .. }) => *value as u64,
+            Item::MetadataEntry(value, _, _) => *value as u64,
```

Comment:
Not blocking for this PR, but I' just wondering - since we have safe wrappers for the most of these types, couldn't we make them hashable and use these hashes instead of memory addresses?

Also, nit:

```suggestion
            Item::GlobalVariable(value)
            | Item::GlobalAlias(value)
            | Item::Function(value)
            | Item::FunctionParam(value)
            | Item::Instruction(value)
            | Item::Operand(Operand { value, .. })
            | Item::MetadataEntry(value, _, _) => *value as u64,

```

Reply:
Consider:

```
let value = <x>;
let i1 = Item::Operand(Operand { ..., value, 1 });
let i2 = Item::Operand(Operand { ..., value, 2});
```

these are two different items, but same value. So we can't hash Item since it would be weird if eg we made the two items above hash to the same thing.

The thing we should do is hash `Value`. However since `Value` has a lifetime and `DISanitizer` doesn't yet, it's not easy. I've considered doing it, but to do it right we need a wrapper type for `LLVMContext` and then make everything borrow from it. It's a large change that is orthogonal to what's being done in this PR.

Reply:
OK, makes sense.